### PR TITLE
make stop property functional

### PR DESCRIPTION
**Pull request by mattf**
`ChatNVIDIA(..., stop=...)` would not change the `stop` words for `invoke`

fixed `_generate` logic and added tests for `stop` across `invoke` and `stream`

future - we should deprecate the `stop` property on `ChatNVIDIA`

### CODE REVIEW

File: libs/ai-endpoints/tests/integration_tests/test_chat_models.py [hunk 1]
```diff
@@ -202,9 +202,23 @@ def test_ai_endpoints_invoke(chat_model: str, mode: dict) -> None:
     assert isinstance(result.content, str)
 
 
-# todo: tests for parameters -
-#        bad: str - Bad words to avoid (cased)
-#        stop: str - Stop words (cased)
+# todo: test that stop is cased and works with multiple words
```
Comment (dglogo) [inline libs/ai-endpoints/tests/integration_tests/test_chat_models.py:205]: from openAI API spec:
"Up to 4 sequences where the API will stop generating further tokens."

File: libs/ai-endpoints/langchain_nvidia_ai_endpoints/chat_models.py [hunk 2]
```diff
@@ -318,7 +318,7 @@ def _get_generation(
         **kwargs: Any,
     ) -> dict:
         """Call to client generate method with call scope"""
-        kwargs["stop"] = kwargs.get("stop", self.stop)
```
Comment (raspawar) [inline libs/ai-endpoints/langchain_nvidia_ai_endpoints/chat_models.py:321]: why this change?

File: libs/ai-endpoints/tests/unit_tests/test_stop.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+    client = ChatNVIDIA(model="mock-model", api_key="mocked", stop=["GO"])
+    response = client.stream("Ok?", stop=expected_stop)
+
+    assert next(response).content == "Ok"
+
+    assert requests_mock.last_request is not None
+    request_payload = requests_mock.last_request.json()
+    assert "stop" in request_payload
+    assert request_payload["stop"] == expected_stop
```
Comment (raspawar) [inline libs/ai-endpoints/tests/unit_tests/test_stop.py:124]: I have a suggestion for the repeating test cases with invoke/stream, these can be combined with [@pytest.mark.parametrize](https://docs.pytest.org/en/7.1.x/how-to/parametrize.html) and stop can be passed as parameter.

**Review by raspawar — CHANGES_REQUESTED** [review]

File: libs/ai-endpoints/langchain_nvidia_ai_endpoints/chat_models.py (hunk 2 above)
Comment (mattf) [reply libs/ai-endpoints/langchain_nvidia_ai_endpoints/chat_models.py:321]: `_generate` has `stop: Optional[List[str]] = None` and always calls `_get_generation(..., stop=stop, ...)`

this means `"stop"` is always in `kwargs`, even if the value is `None`. as a result, `kwags.get("stop", self.stop)` never returns the `self.stop` default.

to get the default of `self.stop` we need to see if `stop` is `not None`

File: libs/ai-endpoints/tests/unit_tests/test_stop.py (hunk 3 above)
Comment (mattf) [reply libs/ai-endpoints/tests/unit_tests/test_stop.py:124]: will you give me an example?

File: libs/ai-endpoints/langchain_nvidia_ai_endpoints/chat_models.py (hunk 2 above)
Comment (raspawar) [reply libs/ai-endpoints/langchain_nvidia_ai_endpoints/chat_models.py:321]: Yeah, makes sense

File: libs/ai-endpoints/tests/unit_tests/test_stop.py (hunk 3 above)
Comment (raspawar) [reply libs/ai-endpoints/tests/unit_tests/test_stop.py:124]: Sure,

this 2 can be combined as

def test_stop_property_invoke(requests_mock: Mocker) -> None:
    expected_stop = ["STOP"]

    client = ChatNVIDIA(model="mock-model", api_key="mocked", stop=expected_stop)
    response = client.invoke("Ok?")

    assert response.content == "Ok"

    assert requests_mock.last_request is not None
    request_payload = requests_mock.last_request.json()
    assert "stop" in request_payload
    assert request_payload["stop"] == expected_stop


def test_stop_parameter_invoke(requests_mock: Mocker) -> None:
    expected_stop = ["STOP"]

    client = ChatNVIDIA(model="mock-model", api_key="mocked")
    response = client.invoke("Ok?", stop=expected_stop)

    assert response.content == "Ok"

    assert requests_mock.last_request is not None
    request_payload = requests_mock.last_request.json()
    assert "stop" in request_payload
    assert request_payload["stop"] == expected_stop

======================================================================
@pytest.mark.parametrize("client_stop, expected_stop", [(None, ["STOP"]), (["GO"], ["STOP"])])
def test_stop_invoke(requests_mock: Mocker, client_stop, expected_stop) -> None:
    client = ChatNVIDIA(model="mock-model", api_key="mocked", stop=client_stop)
    response = client.invoke("Ok?", stop=expected_stop)

    assert response.content == "Ok"

    assert requests_mock.last_request is not None
    request_payload = requests_mock.last_request.json()
    assert "stop" in request_payload
    assert request_payload["stop"] == expected_stop

File: libs/ai-endpoints/langchain_nvidia_ai_endpoints/chat_models.py (hunk 2 above)
Comment (mattf) [reply libs/ai-endpoints/langchain_nvidia_ai_endpoints/chat_models.py:321]: we could pass `stop=stop or self.stop` from `_generate`. i originally went with that before settling on this because it is consistent w/ how the stream and async methods work.

we should really mark the stop property as deprecated.

File: libs/ai-endpoints/tests/unit_tests/test_stop.py (hunk 3 above)
Comment (mattf) [reply libs/ai-endpoints/tests/unit_tests/test_stop.py:124]: nice idea. i'll update to use this.

Comment (mattf) [reply libs/ai-endpoints/tests/unit_tests/test_stop.py]: ptal

**Review by raspawar — APPROVED** [review]
This looks Great!!
